framework/src/auxkernels/TimeDerivativeAux.C

```diff
@@ -0,0 +1,75 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "TimeDerivativeAux.h"
+
+registerMooseObject("MooseApp", TimeDerivativeAux);
+registerMooseObject("MooseApp", ADTimeDerivativeAux);
+
+template <bool is_ad>
+InputParameters
+TimeDerivativeAuxTempl<is_ad>::validParams()
+{
+  InputParameters params = AuxKernel::validParams();
+  params.addClassDescription(
+      "Returns the time derivative of the specified variable/functor as an auxiliary variable.");
+
+  params.addCoupledVar("v", "Variable to take the time derivative of");
+  params.addParam<MooseFunctorName>("functor", "Functor to take the time derivative of");
+  params.addParam<MooseFunctorName>("factor", 1, "Factor to multiply the time derivative by");
+
+  return params;
+}
+
+template <bool is_ad>
+TimeDerivativeAuxTempl<is_ad>::TimeDerivativeAuxTempl(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _v_dot(isParamValid("v") ? &coupledDot("v") : nullptr),
+    _functor(isParamValid("functor") ? &getFunctor<GenericReal<is_ad>>("functor") : nullptr),
+    _factor(getFunctor<GenericReal<is_ad>>("factor")),
+    _use_qp_arg(dynamic_cast<MooseVariableFE<Real> *>(&_var))
+{
+  if ((!_functor && !_v_dot) || (_functor && _v_dot))
+    mooseError("A variable or a functor parameter should be provided to the "
+               "TimeDerivativeAuxTempl<is_ad>");
+
+  if (!_use_qp_arg && !dynamic_cast<MooseVariableFV<Real> *>(&_var))
+    paramError(
+        "variable",
+        "The variable must be a non-vector, non-array finite-volume/finite-element variable.");
+
+  if (isNodal())
+    paramError("variable", "This AuxKernel only supports Elemental fields");
+}
+
+template <bool is_ad>
+Real
+TimeDerivativeAuxTempl<is_ad>::computeValue()
+{
+  using MetaPhysicL::raw_value;
+
+  if (_use_qp_arg)
+  {
+    const auto qp_arg = std::make_tuple(_current_elem, _qp, _qrule);
+
+    if (_v_dot)
+      return raw_value(_factor(qp_arg)) * (*_v_dot)[_qp];
+    else
+      return raw_value(_factor(qp_arg)) * raw_value((*_functor).dot(qp_arg));
+  }
+  else
+  {
+    const auto elem_arg = makeElemArg(_current_elem);
+
+    if (_v_dot)
+      return raw_value(_factor(elem_arg)) * (*_v_dot)[_qp];
```

Comment:
This seems a little awkward since we are in the `!_use_qp_arg` branch. Would it work to call `v_var.dot(elem_arg)` instead?

Reply:
I dont think the dot is implemented for FE variables rn.
I could try to do that though, likely isnt too hard

Reply:
`_use_qp_arg` is always true for FE variables right? Or I guess that you're imagining the aux var could be FV and the functor could be an FE variable? I actually have `evaluateDot` with qp arguments implemented for all variables:
```c++
template <typename OutputType>
typename MooseVariableField<OutputType>::DotType
MooseVariableField<OutputType>::evaluateDot(const ElemQpArg & elem_qp,
                                            const unsigned int state) const
{
  mooseAssert(_time_integrator && _time_integrator->dt(),
              "A time derivative is being requested but we do not have a time integrator so we'll "
              "have no idea how to compute it");
  evaluateOnElement(elem_qp, state);
  const auto qp = std::get<1>(elem_qp);
  mooseAssert(qp < _current_elem_qp_functor_dot.size(),
              "The requested " << qp << " is outside our dot size");
  return _current_elem_qp_functor_dot[qp];
}

template <typename OutputType>
void
MooseVariableField<OutputType>::evaluateOnElementSide(const ElemSideQpArg & elem_side_qp,
                                                      const unsigned int state) const
{
```
The issue is that the element-based arguments aren't implemented for FE

Reply:
well except we're already in the "not a FE variable" part of the loop. So yeah we can use v_var, it ll be a finite volume variable.

Reply:
We're in the `_var` is not an FE variable part of the loop. We have not done any checks on `v_var`

Reply:
Ideally we don't see mixing of `elem_arg` and `_qp` on the same line

Reply:
should be unmixed now.
Would be someone using a FE variable time derivative to fill a FV variable, wont happen for a while haha

I added a rider commit with a warning if you dont mind.

Reply:
When the aux system gets involved, I feel like an accidental mixing of FE and FV is more likely since a user may just copy stuff like
```c++
[AuxVariables]
  [aux]
    family = MONOMIAL
    order = CONSTANT
  []
[]
```
But yes, I agree that someone will not be doing this *on purpose* for quite a while.

However, I think your new error in `Coupleable.h` helps highlight that actually the *only* way we can get to this line of code is for the aux variable to be FV and the coupled variable to be FE ... right? Because `coupledDot` will only return without erroring if the coupled variable is an FE variable. This line is currently uncovered in testing; actually the whole FV branch is uncovered because the `test_fv.i` has its aux variable as FE (some more of that mixing :smile: ).

My suggestion would be that we do not populate `_v_dot` in the initialization list (well populate it, but with `nullptr`), and instead only call `coupledDot` if `_v_var` dynamic casts to `MooseVariableFE<Real>` (or `MooseVariable` for short) and `_use_qp_arg`. Then in the FV branch of `computeValue`, change `if (_v_dot)` to `if (_v_var)`.

This change would allow a user, in an all FV simulation, to get the results they want, instead of an error, if they set the parameter `v` instead of `functor`.

Reply:
Well I dont know what the solution is here. I added some warnings, for the `v` case. I would much rather we delete that parameter tbh, but that would make an object that is unfit for what the discussions user I m trying to help is doing. 

I reworked this a little to add warnings with the implicit conversions that we are doing. This is of sort of limited, to the v parameter case, because when we input a functor:
- we dont know if it s a fv or fe variable (we could with some more code I guess, getFieldVar("functor"), catch it when it s not even variable etc)
- a power user could be trying to do these conversions. This will ultimately be part of the purpose of the functor system, enough flexibility for both FE and FV
 
so to summarize, I dont catch this with a warning:
```
[AuxVariables]
  [aux]
    family = MONOMIAL
    order = CONSTANT
  []
[]

[AuxKernels]
  [deriv]
    type = TimeDerivativeAux
    variable = aux
    functor = some_fv_variable
  []
[]
```
OR this
```
[AuxVariables]
  [aux_fv]
    family = MONOMIAL
    order = CONSTANT
    fv = true
  []
[]

[AuxKernels]
  [deriv]
    type = TimeDerivativeAux
    variable = aux_fv
    functor = some_fe_variable
  []
[]
```

Reply:
I think this extensive checking on FE-FV conversion will have to move up to AuxKernel.C, it s getting too big

Reply:
>  I would much rather we delete that parameter tbh, but that would make an object that is unfit for what the discussions user I m trying to help is doing.

Why is that? It seems like all the dancing that is being done in here would be removed if we just did a functor evaluation with a quadrature point based argument. That should work in all cases because the qp-based functor evaluations are implemented in `MooseVariableField`. I thought maybe you wanted to incorporate `coupledDot` for speed reasons?

Anyway your most recent changes looked fine to me, other than just changing that last `if (_v_dot)` comparison to `if (_v_var)`

Reply:
not really. I was getting an error on INITIAL, thought it was a general issue with how I did things and moved to coupledDot because that was working and is the vetted way of doing this. But qp functor will do. 

My most recent changes are all gone. And so is v_dot :O